### PR TITLE
YAML Defined Tests Scorecard Plugin

### PR DESCRIPTION
**Pull request by AlexNPavel**
**Description of the change:** Add proposal for a YAML Defined Tests plugin for the scorecard

**Motivation for the change:** This was initially proposed in PR #1049, but has since been split off into this PR as the goals of that proposal changed. There is a lot of discussion over this design on that PR.

This PR will initially be created as a draft as I have not yet implemented various improvements that I've planned. It will be changed to a standard PR once I make those changes.

This PR also depends on PR #1049 getting merged first, and it may be wise to wait for the implementation of that proposal before fully moving forward with this PR.

### CODE REVIEW

**Comment by AlexNPavel** [comment]
This is ready for review.

File: doc/proposals/yaml-scorecard-plugin.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    Spec map[string]interface{} `mapstructure:"spec"`
+    // Expected resources and status
+    Expected Expected `mapstructure:"expected"`
+}
+```
+
+For `Status` and `Resources` fields, we can implement a bit of extra computation instead of simple string checking. For instance,
+in the memcached-operator test, we should expect that the length of the `nodes` field (which is an array) has a certain length. To implement functions like
+these, we can create some functions for these checks that are prepended by `scorecard_function_` and take an array of objects. For instance, in the above
+example, `scorecard_function_length` would check that each field listed under it matches the specified length (like `nodes: 4`). If the yaml key does not
```
Comment (estroz) [inline doc/proposals/yaml-scorecard-plugin.md]: Can you give an example of how to implement a `scorecard_function_`?

Comment (AlexNPavel) [reply doc/proposals/yaml-scorecard-plugin.md]: The scorecard functions are simply a part of the main yaml parser and just change the way we handle a submap. I have the `scorecard_function_length` in my old implementation.

Parser to get the interfaces we are checking: https://github.com/AlexNPavel/operator-sdk/blob/scorecard-usertest//commands/operator-scorecard/lib/lib.go#L284:L296
Length function: https://github.com/AlexNPavel/operator-sdk/blob/scorecard-usertest//commands/operator-scorecard/lib/lib.go#L198:L213

The only functions I can think of that we would need are array length and regex

Comment (estroz) [reply doc/proposals/yaml-scorecard-plugin.md]: Ok it would be good to at least list those two in this section.

**Comment by estroz** [comment]
@jmccormick2001 is this doc still relevant after the changes you've made? If so we should probs get this merged @AlexNPavel.

**Comment by jmccormick2001** [comment]
> @jmccormick2001 is this doc still relevant after the changes you've made? If so we should probs get this merged @AlexNPavel.

I'd probably wait to merge it since we have just entered new stories to define a new 'custom test' design/interface that would work with the other scorecard changes in the queue all as part of v1alpha2 API changes.

**Comment by camilamacedo86** [comment]
Why @estroz /hold this one? What is missing for we merge this?

**Pull request closed without merging**